tests/macsec/__init__.py

```diff
@@ -0,0 +1,181 @@
+import pytest
+import os
+import natsort
+import json
+import collections
+from ipaddress import ip_address, IPv4Address
+
+from macsec_config_helper import enable_macsec_feature
+from macsec_config_helper import disable_macsec_feature
+from macsec_config_helper import setup_macsec_configuration
+from macsec_config_helper import cleanup_macsec_configuration
+
+logger = logging.getLogger(__name__)
+with open(os.path.dirname(__file__) + '/profile.json') as f:
+    macsec_profiles = json.load(f)
+
+
+def get_macsec_profile_list():
+    return natsort.natsorted(macsec_profiles.keys())
+
+
+def pytest_addoption(parser):
+    parser.addoption("--enable_macsec", action="store_true", default=False,
+                     help="Enable macsec on some links of testbed")
+    parser.addoption("--macsec_profile", action="store", default="all",
+                     type=str, help="profile name list in macsec/profile.json")
+
+
+def pytest_configure(config):
+    config.addinivalue_line(
+        "markers", "macsec_required: mark test as MACsec required to run")
+
+
+def pytest_collection_modifyitems(config, items):
+    if not config.getoption("enable_macsec"):
+        skip_macsec = pytest.mark.skip(reason="macsec test cases")
+        for item in items:
+            if "macsec_required" in item.keywords:
+                item.add_marker(skip_macsec)
+
+
+def pytest_generate_tests(metafunc):
+    if 'macsec_profile' in metafunc.fixturenames:
+        if metafunc.config.getoption("enable_macsec"):
+            profiles = _generate_macsec_profile(metafunc)
+            assert profiles, "Specify valid macsec profile!"
+        else:
+            profiles = ['']
+        metafunc.parametrize('macsec_profile', profiles, scope="session")
+
+
+def _generate_macsec_profile(metafunc):
+    value = metafunc.config.getoption("macsec_profile")
+    if value == 'all':
+        return get_macsec_profile_list()
+    return [x for x in value.split(',') if x in macsec_profiles]
+
+
+@pytest.fixture(scope="session")
+def macsec_feature(request, duthost, macsec_nbrhosts):
+    if request.config.getoption("enable_macsec"):
+        enable_macsec_feature(duthost, macsec_nbrhosts)
+        yield
+        disable_macsec_feature(duthost, macsec_nbrhosts)
+    else:
+        yield
+        return
```

Comment:
Why is this `else` necessary?

Reply:
It's a generator fixture, `yield` statement is needed in `else` too.